### ISSUE TITLE
New rule: SC4-1-1-unique-attrs

### ISSUE THREAD

**Issue by WilcoFiers** [auto-wcag/auto-wcag#328]
---
name: attributes are not duplicated
rule_type: atomic
description: |
  This rule checks that HTML and SVG starting tags do not contain duplicated attributes.

success_criterion:
- 4.1.1 # Parsing

test_aspects:
- Source code

authors:
- Wilco Fiers
- Emma Pratt Richens
---

## Applicability

The rule applies to any [starting tag](https://www.w3.org/TR/html5/syntax.html#start-tags) in an HTML or SVG file.

**Note:** This rule cannot be tested on the DOM Tree because the browser ignores duplicated attributes.

## Expectation

For each test target, there are no duplicated [attributes](https://www.w3.org/TR/html5/syntax.html#elements-attributes).

## Assumptions

*There are currently no assumptions.*

## Accessibility support

*There are no major accessibility support issues known for this rule.*

## Background

- - [H94: Ensuring that elements do not contain duplicate attributes](https://www.w3.org/TR/WCAG20-TECHS/H94.html)

#### Passed example 1

No attributes are duplicated.

```html
<img src="image.jpg" alt="" />
```

#### Passed example 2

No attributes, therefore no attributes are duplicated.

```html
<br />
```

#### Passed example 3

Empty attributes, no attributes are duplicated.

```html
<input type="checkbox" disabled readonly />
```

#### Passed example 4

SVG, no attributes are duplicated.

```html
<svg>
  <line x1="0" y1="0" x2="200" y2="200" style="stroke-width:2" />
</svg>
```

#### Passed example 5

Script, no attributes are duplicated. HTML or SVG code within a script should be ignored.

```html
<script>
	var foo = '<img src="image.jpg" alt="" alt="image" />'
</script>
```

#### Failed example 1

At least one attribute is duplicated.

```html
<img src="image.jpg" alt="" alt="image" />
```

#### Failed example 2

Empty attributes, at least one attribute is duplicated.

```html
<input type="checkbox" disabled="disabled" disabled readonly />
```

#### Failed example 3

SVG, at least one attribute is duplicated.

```html
<svg>
  <line x1="0" y1="0" x1="200" y1="200" style="stroke-width:2" />
</svg>
```

#### Inapplicable example 1

Code is XML, and not an HTML or SVG document.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<earl:TestResult rdf:about="#result"></earl>
```

#### Inapplicable example 2

Code is JavaScript, and not an HTML or SVG document.

```js
var foo = '<img src="image.jpg" alt="" alt="image" />'
```

**Comment by WilcoFiers** [comment]
## Source code

A set of instructions and statements to be interpreted by a computer. Each source code file has a file type. The method for determining file type depends on the method through which the file is loaded.

**Comment by EmmaJP** [comment]
Resolved by pull request #332
